safe-transaction-client/src/lib.rs

```diff
@@ -0,0 +1,343 @@
+        } = signed_safe_tx.inner;
+        self.client
+            .agent
+            .post(&format!(
+                "{}/v1/safes/{}/multisig-transactions/",
+                self.client.transactions_api,
+                to_checksum(&self.safe_address, None),
+            ))
+            .send_json(ureq::json!({
+                "to": to_checksum(&to, None),
+                "value": format!("{}", value),
+                "data": data,
+                "operation": operation as u8,
+                "gasToken": None::<()>,
+                "safeTxGas": "0",
+                "baseGas": "0",
+                "gasPrice": "0",
+                "refundReceiver": Address::zero(),
+                "nonce": format!("{}", nonce),
+                "contractTransactionHash": signed_safe_tx.safe_tx_hash,
+                "sender": to_checksum(&signed_safe_tx.sender, None),
+                "signature": format!("0x{}", signed_safe_tx.signature),
+                "origin": "An unauthenticated origin string?",
+            }))
+            // 200 OK, is not okay. We expect 201 Created.
+            .and_then(|r| match r.status() {
+                201 => Ok(()),
+                s => Err(ureq::Error::Status(s, r)),
+            })
+            .map_err(From::from)
+    }
+
+    pub fn add_to_confirmations(&self, signed_safe_tx: SignedSafeTxHash) -> Result {
+        self.client
+            .agent
+            .post(&format!(
+                "{}/v1/multisig-transactions/{:?}/confirmations/",
+                self.client.transactions_api, signed_safe_tx.safe_tx_hash,
+            ))
+            .send_json(ureq::json!({
+                "signature": format!("0x{}", signed_safe_tx.signature),
+            }))
+            // 200 OK, is not okay. We expect 201 Created.
+            .and_then(|r| match r.status() {
+                201 => Ok(()),
+                s => Err(ureq::Error::Status(s, r)),
+            })
+            .map_err(From::from)
+    }
+}
+
+impl SafeTx {
+    // We consume self. If signing fails, we have bigger issues than recreating the transaction.
+    pub async fn sign<S>(self, signer: &S) -> Result<SignedSafeTx, Error<S::Error>>
+    where
+        S: Signer,
+        Error<S::Error>: From<S::Error>,
+    {
+        use ethers::abi::Tokenizable;
+        use ethers::utils::keccak256;
+        use tiny_keccak::{Hasher, Keccak};
+        macro_rules! tokenize {
+            ($($tok:expr),* $(,)*) => {
+                &[$($tok.into_token()),*][..]
+            }
+        }
+
+        let safe_tx_hash = TxHash::from({
+            let mut hasher = Keccak::v256();
+            let mut output = [0u8; 32];
+            hasher.update(&[0x19, 0x01]);
+            // domain separator
+            hasher.update(&keccak256(&ethers::abi::encode(tokenize![
+                // DOMAIN_SEPARATOR_TYPEHASH
+                U256::from_str(
+                    "0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218"
+                )
```

Comment:
Would be nice if these were statically defined at the top of the file, but `from_str` is probably not `const`..

Reply:
`core::convert::FromStr::from_str` is indeed not `const`. However... `U256([u64; 4])` would work, but would not be as trivial to compare against the contract as a copy-paste search. Alternatively the string itself could be the constant.

I prefer the locality for these single-use constants.

Which would you prefer?

Reply:
yeah... let's keep it like this I guess.